keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,973 @@
+-->
+
+#### Discussions
+
+The Batch WG incubates a new KEP [Backoff Limit Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/issues/3850),
+which introduces a new field `backoffLimitPerIndex`, then each pod of the Job will have an independent backOffLimit,
+but initially it only supports `restartPolicy=Never`, maybe we can widen the _max_restart_times_ to support other restart policies,
+like `Never`, then the Job API no longer needs to consider these logics anymore. Currently, it uses an annotation `batch.kubernetes.io/job-index-failure-count`
+to record the retires number, which looks not elegant.
+
+Pros:
+  * Reduce the maintenance cost of Job API
+
+Cons:
+  * The boundary of restartPolicy `Never` and `OnFailure` is vaguely.
+
+#### Caveats
+
+- Pod's maxRestartTimesOnFailure is somehow conflict with Job's backoffLimit, we'll respect the first one who hit the line.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- If `maxRestartTimesOnFailure` is configured in a recoverable circumstance with a small number retries.
+E.g. we set the `maxRestartTimesOnFailure=3`, and the network shakes with 4 times, the service will stuck
+in a failed state forever after 3 restarts, so end-users should be careful when configuring this field.
+However, this problem also exists in Job's backoffLimit, which is default to 6.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Add `maxRestartTimesOnFailure` to `pod.spec`:
+
+```golang
+type PodSpec struct {
+  // ...
+  RestartPolicy RestartPolicy
+  // maxRestartTimesOnFailure indicates the maximum count container can restart. It's the sum of
+  // all containers' restart times and only works when RestartPolicy is set to `OnFailure`.
+  // Default to nil which means will restart infinity.
+  // +optional
+  maxRestartTimesOnFailure *int
+}
+```
+
+- If Pod's `RestartPolicy` is `Always` or `Never`, `maxRestartTimesOnFailure` is default to nil, and will not apply.
+- If Pod's `RestartPolicy` is `OnFailure`, `maxRestartTimesOnFailure` is also default to nil, which means infinite
+restart times for backwards compatibility.
+
+In runtime, we'll check the sum of `RestartCount` of
+all containers [`Status`](https://github.com/kubernetes/kubernetes/blob/451e1fa8bcff38b87504eebd414948e505526d01/pkg/kubelet/container/runtime.go#L306-L335)
```

Comment:
Summing up all the container-restarts is...weird.

It feels like this should be a per-container setting rather than per-pod.

Reply:
That's a UX thing.  The API for restart counting is actually per-container (in ContainerStaus).

Reply:
If I may add a reason for doing this at the container level (as opposed to taking the sum of the restarts).

I would expect some containers in a Pod spec to be restarted nominally or more frequently than others. For example, it may be desirable that an `initContainer`  waiting for some dependency to become available, be allowed to restart an infinite amount of times, but the main application container itself be limited to only n restarts.

Reply:
Make sense.

Reply:
> The API for restart counting is actually per-container (in ContainerStaus).

Yes, but then it should named like `maxRestartsPreContainer` or under the container field. Now, if a container restarts, then we think a pod restarts once.

Reply:
> it may be desirable that an initContainer waiting for some dependency to become available, be allowed to restart an infinite amount of times, but the main application container itself be limited to only n restarts.

Great inputs. But even initContainer will restart infinitely because of unrecoverable reasons. Anyway, need to think twice.

Reply:
The fact that we show it in `kubectl` as pod restarts is a lie.  They are container restarts.

> it should named like `maxRestartsPreContainer`

`Pod.spec.containers[].maxRestarts` reads well to me.

Reply:
> `Pod.spec.containers[].maxRestarts` reads well to me.

But it's a little weird because Containers also have a `RestartPolicy` and the only allowed value is `Always`. It would stop being intuitive how the attribute interactions actually work because now we're intermingling pod level attributes with container level attributes.